domain/removal/state/model/operation.go

```diff
@@ -0,0 +1,245 @@
+// getEmptyOperationUUIDs returns the UUIDs of operations that are empty.
+func (st *State) getEmptyOperationUUIDs(ctx context.Context, tx *sqlair.TX, operationUUIDs []string) ([]string, error) {
+	if len(operationUUIDs) == 0 {
+		return nil, nil
+	}
+	type operation entityUUID
+
+	stmt, err := st.Prepare(`
+SELECT DISTINCT o.uuid AS &operation.uuid
+FROM operation AS o
+LEFT JOIN operation_task AS ot ON o.uuid = ot.operation_uuid 
+WHERE o.uuid IN ($uuids[:])
+AND ot.uuid IS NULL`, operation{}, uuids{})
+	if err != nil {
+		return nil, errors.Capture(err)
+	}
+
+	var result []operation
+	if err := tx.Query(ctx, stmt, uuids(operationUUIDs)).GetAll(&result); err != nil && !errors.Is(err, sqlair.ErrNoRows) {
+		return nil, errors.Errorf("running get empty operation query: %w", err)
+	}
+
+	return transform.Slice(result, func(op operation) string {
+		return op.UUID
+	}), nil
+}
+
+// deleteTaskByUUIDs deletes the tasks and their associated dependencies by
+// their UUIDs.
+// It returns the storeUUIDs of operation_task_output to allow cleanup
+// of the object datastore in the service layer.
+func (st *State) deleteTaskByUUIDs(ctx context.Context, tx *sqlair.TX, toDelete []string) ([]string, error) {
+	if len(toDelete) == 0 {
+		return nil, nil
+	}
+
+	// get store paths
+	type store struct {
+		Path string `db:"path"`
+	}
+	tasks := uuids(toDelete)
+	stmt, err := st.Prepare(`
+SELECT &store.path
+FROM object_store_metadata_path AS osp
+JOIN operation_task_output AS oto ON osp.metadata_uuid = oto.store_uuid
+WHERE task_uuid IN ($uuids[:])`, tasks, store{})
+	if err != nil {
+		return nil, errors.Capture(err)
+	}
+	var stores []store
+	if err := tx.Query(ctx, stmt, tasks).GetAll(&stores); err != nil && !errors.Is(err, sqlair.ErrNoRows) {
+		return nil, errors.Errorf("getting task store uuids: %w", err)
+	}
+
+	// Delete the tasks dependencies.
+	for _, table := range []string{
+		"operation_unit_task",
+		"operation_machine_task",
+		"operation_task_output",
+		"operation_task_status",
+		"operation_task_log",
+	} {
+		if err := st.deleteByValues(ctx, tx, table, "task_uuid", toDelete); err != nil {
+			return nil, errors.Errorf("deleting %s by task UUIDs: %w", table, err)
+		}
+	}
+
+	// delete the tasks
+	if err := st.deleteByValues(ctx, tx, "operation_task", "uuid", toDelete); err != nil {
+		return nil, errors.Errorf("deleting task by UUIDs: %w", err)
+	}
+
+	return transform.Slice(stores, func(v store) string {
+		return v.Path
+	}), nil
+}
+
+// deleteOperationByUUIDs deletes the operations and their associated dependencies by
+// their UUIDs.
+func (st *State) deleteOperationByUUIDs(ctx context.Context, tx *sqlair.TX, toDelete []string) error {
+	// Delete the operations dependencies.
+	for _, table := range []string{
+		"operation_action",
+		"operation_parameter",
+	} {
+		if err := st.deleteByValues(ctx, tx, table, "operation_uuid", toDelete); err != nil {
+			return errors.Errorf("deleting %s by operation UUIDs: %w", table, err)
+		}
+	}
```

Comment:
todo: The comment is misleading and needs an updating. `operation_action` is not a dependency of all operations. We're deleting from a table with no references in the case of `juju exec` type operations. "Luckily" delete and update do not return `sql.ErrNoRows` if not changes were made, nor rows deleted.